packages/flutter/lib/src/widgets/selection_gestures.dart

```diff
@@ -0,0 +1,1320 @@
+          break;
+        }
+        if (pastTapTolerance) {
+          // This means our pointer was not accepted as a tap nor a drag.
+          // This can happen when a user drags on a right click, going past the
+          // tap tolerance, and drag tolerance, but being rejected since a right click
+          // drag is not allowed by this recognizer.
+          _checkCancel();
+          resolve(GestureDisposition.rejected);
+        } else {
+          _checkDragCancel();
+          if (currentUp != null) {
+            _checkTapUp(currentUp!);
+          }
+        }
+        break;
+
+      case _DragState.accepted:
+        // We only arrive here, after the recognizer has accepted the [PointerEvent]
+        // as a drag. Meaning [_checkTapDown], and [_checkDragStart] have already ran.
+        _checkDragEnd();
+        _initialButtons = null;
+        break;
+    }
+
+    _stopDeadlineTimer();
+    _dragState = _DragState.ready;
+  }
+
+  @override
+  void handleEvent(PointerEvent event) {
+    super.handleEvent(event);
+    if (event is PointerMoveEvent) {
+      // Receiving a [PointerMoveEvent], does not automatically mean the pointer
+      // being tracked is doing a drag gesture. There is some drift that can happen
+      // between the initial [PointerDownEvent] and subsequent [PointerMoveEvent]s,
+      // that drift is handled by the tap status tracker. Accessing [_TapStatusTracker.pastTapTolerance]
+      // lets us know if our tap has moved past the acceptable tolerance. If the pointer
+      // does not move past this tolerance than it is not considered a drag.
+      //
+      // To be recognized as a drag, the [PointerMoveEvent] must also have moved
+      // a sufficient global distance from the initial [PointerDownEvent] to be
+      // accepted as a drag. This logic is handled in [_hasSufficientGlobalDistanceToAccept].
+
+      // If the buttons differ from the [PointerDownEvent]s buttons then we should stop tracking
+      // the pointer.
+      if (event.buttons != _initialButtons) {
+        _giveUpPointer(event.pointer);
+      }
+
+      if (_dragState == _DragState.accepted) {
+        _checkDragUpdate(event);
+      } else if (_dragState == _DragState.possible) {
+        _checkDrag(event);
+
+        // We may arrive here if the recognizer is accepted before a [PointerMoveEvent] has been
+        // received.
+        if (_start != null && _wonArenaForPrimaryPointer) {
+          _acceptDrag(_start!);
+        }
+      }
+    } else if (event is PointerUpEvent) {
+      if (_dragState == _DragState.possible) {
+        // If we arrive at a [PointerUpEvent], and the recognizer has not won the arena, and the tap drift
+        // has exceeded its tolerance, then we should reject this recognizer.
+        if (pastTapTolerance) {
+          _giveUpPointer(event.pointer);
+          return;
+        }
+        // The drag has not been accepted before a [PointerUpEvent], therefore the recognizer
```

Comment:
I can't wrap my head around the logic.
Is it using `pastTapTolerance` to infer the drag has not been accepted?

Reply:
Say we have a sequence of events `PointerDownEvent` -> `PointerMoveEvent` -> `PointerUpEvent`. There are 3 cases.

If we arrive at that `PointerUpEvent` and the `PointerMoveEvent` did not move a sufficient global distance to be accepted as a drag, but moved a sufficient global distance to move past the tap tolerance.
- TapCancel
- DragCancel

If we arrive at that `PointerUpEvent` and the `PointerMoveEvent` moved a sufficient global distance to be accepted as a drag, therefore also moving past the tap tolerance because dragTolerance > tapTolerance.
- TapCancel

If we arrive at that `PointerUpEvent` and the `PointerMoveEvent` did not move a sufficient global distance to be accepted as a drag and it did not move past the tap tolerance.
- DragCancel

Reply:
> the pointer did not move a sufficient global distance to be accepted as a drag.

In this case, why does it need to send DragCancel? Does the Drag already start at this point?

Reply:
It looks like in case #1 I am not matching the target behavior. Where individual `TapGestureRecognizer` and `DragGestureRecognizer` will detect the case as a drag, my single `TapAndDragGestureRecognizer` rejects both tap and drag.

Reply:
I am now mostly matching the behavior in scenario `#1`. 

With two recognizers `TapGestureRecognizer` and `PanGestureRecognizer`:
`onDown-PanGestureRecognizer` -> `onStart-PanGestureRecognizer` -> `onUpdate-PanGestureRecognizer` -> `onEnd-PanGestureRecognizer`

With one `TapAndDragGestureRecognizer`:
`onTapDown-TapAndDragGestureRecognize` -> `onTapCancel-TapAndDragGestureRecognize` -> `onDragStart-TapAndDragGestureRecognize` -> `onDragUpdate-TapAndDragGestureRecognize` -> `onDragEnd-TapAndDragGestureRecognize`.

The difference is that the new recognizer calls `onTapDown` and `onTapCancel` followed by `onDragStart` and `onDragEnd`. This differing behavior is expected for the new recognizer since it will call `onTapDown` when `acceptGesture` is called, and that needs to be followed up with an `onTapCancel` when a drag begins. 

When using two recognizers (`TapGestureRecognizer` and `PanGestureRecognizer`), `onTapDown` is not called because `TapGestureRecognizer` only fires it on `acceptGesture` and `TapGestureRecognizer` self declares loss when moving past tap tolerance, therefore there is also no need for it to call `onTapCancel` since tap down was never fired. In this situation `PanGestureRecognizer` wins and the gesture is pan.

Our case differs because `TapGestureRecognizer` declares loss, which gives `acceptGesture` in `TapAndDragGestureRecognizer` the chance to call `onTapDown`, but because our pointer moved past the tap tolerance this gesture cannot be a tap, so we call tap cancel and the gesture defaults to a drag.

Reply:
It feels weird that we merge two gesture recognizers to eliminate the conflicting callback and cancelling each other, but now we still simulating the canceling behavior in one merged gesture recognizer. In my opinion, once the canceling is called, this gesture recognizer should not fire any other callback in this gesture sequence. I think we should treat tap and drag as one kind of gesture so in this case 
`onTapDown-TapAndDragGestureRecognize -> onTapCancel-TapAndDragGestureRecognize -> onDragStart-TapAndDragGestureRecognize -> onDragUpdate-TapAndDragGestureRecognize -> onDragEnd-TapAndDragGestureRecognize.`

it doesn't make sense to call `onTapCancel-TapAndDragGestureRecognize`. The `onTapDown` can be a down for a single tap or a down for a drag. The cusumer (i.e editabletext) should have this in mind when using this recognizer. If they have to differentiate it, they need to implement their own logic or keep some kind of state.

Does this make sense?

Reply:
This makes a lot of sense! I think my initial train of thought regarding the decision on cancel behavior was to keep some tests happy, but it makes a lot more sense to change the intended behavior and have the recognizer only have one cancel.

Reply:
Updated the cancel behavior to only cancel when a tap down has been fired, i.e. when a PointerDownEvent was received. Also merged `onTapCancel` and `onDragCancel` into `onCancel`. I agree that the `onCancel` should be the callback fired in a sequence of callbacks of the recognizer.